### ISSUE TITLE
Replication through proxy fails

### ISSUE THREAD

**Issue by cubicleWar** [pouchdb/pouchdb#5678]
### Issue

In many circumstances it is desirable to put couchdb behind a application server so that authentication can be carried out in common with other aspects of the app. When trying to replicate between pouchdb and couchdb sitting behind a php proxy I always receive the following error:

    TypeError: undefined is not an object (evaluating 'changes.results')

I assumed this may be due to long polling of the _changes feed so I tried to disable all those options:

    this._db.sync(remoteDB, {
        live: false,
        timeout: false,
        heartbeat: false,
        filter: '_design/app/by_app',
        query_params: { "app" : "convert" }
    }).on('complete', function () {
        console.log('success');
    }).on('error', function (err) {
        // boo, something went wrong!
        console.log(err);
    }); 

However I noticed that if I curl the requests pouchdb is making like:

    curl http://my.app/test/db/_changes?style=all_docs&since=0&limit=100

It is still using long polling, however if I remove the *since* and *limit* query parameters it is not.

If there a way to support replication through a proxy where long polling/continuous is not supported in pouchdb?

I'm not completely sure if this issue is due to improper use of pouchdb replication, my proxy or a couchdb issue.

When I curl my proxy it seems to return the same as if I curled couchdb directly:

    curl http://my.app/test/db/_changes?style=all_docs&since=0&limit=100
    [1] 57256
    [2] 57257
    [2]+  Done                    since=0
    $ {"results":[
    {"seq":1,"id":"_design/only_correct_user","changes":[{"rev":"1-4d8b2bf4a40888f60b9dd39fbda061ed"}]},
    {"seq":3,"id":"convert_app_settings","changes":[{"rev":"77-dd64223fa38936cfa49bd27df41854a2"}],"deleted":true},
    {"seq":4,"id":"_design/app","changes":[{"rev":"1-6f2f7157229fa647a01acbaad51c7e38"}]},
    {"seq":6,"id":"test","changes":[{"rev":"2-56ca4092d53e33094eea6eb0807a8a07"}],"deleted":true},
    {"seq":7,"id":"d4f8ffedfd7824bf1cecdf1eb700ce74","changes":[{"rev":"1-6eb0fc8f9f1201cde945390a9a0dc4a8"}]}
    ],
    "last_seq":7}

Curl sits there until I hit enter then returns:

    [3] 57281
    [2]   Done                    curl http://my.app/test/db/_changes?style=all_docs

### Info
- Environment: Hybrid app (Ionic 2)
- Platform: iOS simulator
- Server: CouchDB 1.6.0 behind php proxy

Notes 
- I have also attempted to use express-couch-proxy (node) with the same results
- I am relatively new to couchDB so may be missing some simple inputs
- Proxy http://my.app/test to http://my.app:5984

**Comment by willholley** [comment]
@cubicleWar longpoll _changes feeds require an explicit query parameter to Couch (see http://docs.couchdb.org/en/stable/api/database/changes.html#changes). Are you able to capture the complete request/response (e.g. using browser developer tools) for the request that PouchDB is failing on?

**Comment by cubicleWar** [comment]
I logged the url and the couchdb responses from the proxy and this is the result:

    [2016-09-14 01:58:40] local.INFO: GET http://my.app:5984/twal485257d8ac48dbeae/  
    [2016-09-14 01:58:40] local.INFO:      {"db_name":"twal485257d8ac48dbeae","doc_count":1,"doc_del_count":0,"update_seq":1,"purge_seq":0,"compact_running":false,"disk_size":4194,"data_size":506,"instance_start_time":"1473817672927950","disk_format_version":6,"committed_update_seq":1}
  
    [2016-09-14 01:58:40] local.INFO: GET http://my.app:5984/  
    [2016-09-14 01:58:40] local.INFO: {"couchdb":"Welcome","uuid":"50d9237c6e8d069d57f25adcf14885e5","version":"1.6.0","vendor":{"version":"15.10","name":"Ubuntu"}}
  
    [2016-09-14 01:58:40] local.INFO: GET http://my.app:5984/twal485257d8ac48dbeae/_local/Y5pT6HYz5ss8bQxGqNhRgA%3D%3D?  
    [2016-09-14 01:58:40] local.INFO: {"error":"not_found","reason":"missing"}
  
    [2016-09-14 01:58:40] local.INFO: GET http://my.app:5984/twal485257d8ac48dbeae/_changes?style=all_docs&app=convert&since=0&limit=100  
    [2016-09-14 01:58:40] local.INFO: {"results":[
    {"seq":1,"id":"_design/db_access","changes":[{"rev":"1-2f51b6ccac872b9dbf7b9d758f74fc8b"}]}  
    ],
    "last_seq":1}
  
  
    [2016-09-14 01:58:40] local.INFO: GET http://my.app:5984/twal485257d8ac48dbeae/_local/VnnYXnEODBk79aHrlLBpdQ%3D%3D?  
    [2016-09-14 01:58:40] local.INFO: {"error":"not_found","reason":"missing"}
  

This is for a sync to a fresh couchdb installation.

**Comment by cubicleWar** [comment]
I have also tried using pouchdb in the browser to get rid of another complication in debugging and when I look in dev tools it seems like I occasionally get an error trying to load resource for the following:

     http://my.app/test/twal485257d8ac48dbeae/_changes?style=all_docs&app=convert&since=0&limit=100&_nonce=1473837071903

However I get the `TypeError: undefined is not an object (evaluating 'e.results') — pouchdb.min.js:5320` error every time I attempt to sync.

When I hit that URL in the browser I never have any issues and get the following response:

    {"results":[
    {"seq":1,"id":"_design/db_access","changes":[{"rev":"1-2f51b6ccac872b9dbf7b9d758f74fc8b"}]}
    ],
    "last_seq":1}

**Comment by willholley** [comment]
@cubicleWar testing in the browser, are you able to capture the response (using browser dev tools) when the error is thrown?

**Comment by cubicleWar** [comment]
I believe the error is thrown on `http://my.app/test/testuser/_changes?style=all_docs&app=convert&since=0&limit=100&_nonce=1474007668692`

I have attached a sample project with a vagrant server that will let you observe it first hand.

[pouchdb.zip](https://github.com/pouchdb/pouchdb/files/476240/pouchdb.zip)

**Comment by willholley** [comment]
@cubicleWar minor thing but there's an error in after.sh - the testuser document needs to have `"name":"testuser"` to be valid.

There also seems to be a bit of strangeness around HTTP headers / duplication.

Anyway, the root of the problem appears to be the replication filter.  Specifically, you are specifying a filter function that doesn't exist on the client. Sync is bi-directional, so if you specify a filter like this, it needs to exist at both ends. The way PouchDB fails isn't very friendly but you could work around it by doing:

```
localDB.sync(remoteDB, {
        push: {
            live: false,
            timeout: false,
            heartbeat: false
        },
        pull: {
            live: false,
            timeout: false,
            heartbeat: false,
            filter: '_design/db_access/by_app',
            query_params: { "app" : "convert" }
        }
    }).on('complete', function () {
        // yay, we're done!
        console.log('success');
    }).on('error', function (err) {
        // boo, something went wrong!
        console.log(err);
    });
```

**Comment by cubicleWar** [comment]
@willholley That works great however I added the design document to the local db and it seems it still fails if I specify the filter and query_params on push, although I probably need to test a bit more.

In terms of HTTP headers/duplication weirdness is there anything specific I can look into?

**Comment by cubicleWar** [comment]
I have updated the client code to allow adding of the the design document to pouchdb and trying 'symmetric' sync (using a design document filter on both sides). 

What is interesting is it seems to fail when using the filter in the design doc but if I use a filter specified as a parameter it works fine. An example of this is the plain sync() function that now specifies a filter function as part of the push config options.  

I also noticed pouch reporting some 400 bad requests on _local documents however logging on the server side seems to say everything is a 200 or 201.

[client.zip](https://github.com/pouchdb/pouchdb/files/478036/client.zip)

**Comment by willholley** [comment]
@cubicleWar that design doc looks invalid. Instead of:

```
var designDoc = {
	_id : '_design/db_access',
	/*validate_doc_update : 'function (newDoc, oldDoc, userCtx) {if (userCtx.roles.indexOf(\'_admin\') === -1 && userCtx.name != \'' + userName + '\') {throw({forbidden : \'Not Authorised\'});}}',*/
	filters : {
		by_app : "function(doc, req) { return doc.app === req.query.app;}.toString()"
	}
};
```

try:
```
var designDoc = {
	_id : '_design/db_access',
	/*validate_doc_update : 'function (newDoc, oldDoc, userCtx) {if (userCtx.roles.indexOf(\'_admin\') === -1 && userCtx.name != \'' + userName + '\') {throw({forbidden : \'Not Authorised\'});}}',*/
	filters : {
		by_app : "function(doc, req) { return doc.app === req.query.app;}"
	}
};
```

There's no need for the `toString()` if you manually stringify the function.

**Comment by cubicleWar** [comment]
@willholley changing the filter in the design doc does not seem to fix the issue I still get `TypeError: undefined is not an object (evaluating 'changes.results') — pouchdb.js:13038`. 

I also though that having to stringify the function was weird however its whats stated in [this pouch article](https://pouchdb.com/2015/04/05/filtered-replication.html).

It also appears if you change the document on the server our add a new document that satisfies the filter it won't pull those changes even with your originally recommended push/pull config.

**Comment by willholley** [comment]
@cubicleWar in the article you reference, the filter function isn't already in a string, hence the stringify call, e.g.:

```
"function(doc,req) { // do stuff }"
```
vs
```
function(doc,req) { // do stuff }.toString()
```

We have tests that cover client side replication filters so there's something odd going on. I'd suggest testing without your proxy to verify that's not the issue. If you can create a minimal client-only test that demonstrates the error, that would be even better.

**Comment by cubicleWar** [comment]
@willholley of course, I should of paid more attention to the filter function.

I think I have got to the bottom of it, there were a couple of issues:

-  Instead of specifying the filter as `'_design/db_access/by_app'` it should of been  `'db_access/by_app'`.
- My proxy was always returning status 200 even when it shouldn't of been, for example when pouchdb is checking if the _bulk_get api is supported. Once I switched it to return the status and body from couchdb, pouchdb worked fine.

The only other (separate) issue is that pouchdb is typing to replicate _local documents as I am getting a couple of 400, bad request, invalid_json on PUT requests of _local documents
